shell/src/main/java/alluxio/shell/command/CpCommand.java

```diff
@@ -0,0 +1,236 @@
+
+import alluxio.AlluxioURI;
+import alluxio.Configuration;
+import alluxio.client.ReadType;
+import alluxio.client.WriteType;
+import alluxio.client.file.FileInStream;
+import alluxio.client.file.FileOutStream;
+import alluxio.client.file.FileSystem;
+import alluxio.client.file.URIStatus;
+import alluxio.client.file.options.CreateFileOptions;
+import alluxio.client.file.options.OpenFileOptions;
+import alluxio.exception.AlluxioException;
+import alluxio.exception.ExceptionMessage;
+import alluxio.exception.FileDoesNotExistException;
+import alluxio.shell.AlluxioShellUtils;
+import alluxio.util.io.PathUtils;
+
+import com.google.common.base.Joiner;
+import com.google.common.io.Closer;
+import org.apache.commons.cli.CommandLine;
+import org.apache.commons.io.IOUtils;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Copies a file or a directory in the Alluxio filesystem.
+ */
+@ThreadSafe
+public final class CpCommand extends AbstractShellCommand {
+
+  /**
+   * @param conf the configuration for Alluxio
+   * @param fs the filesystem of Alluxio
+   */
+  public CpCommand(Configuration conf, FileSystem fs) {
+    super(conf, fs);
+  }
+
+  @Override
+  public String getCommandName() {
+    return "cp";
+  }
+
+  @Override
+  protected int getNumOfArgs() {
+    return 2;
+  }
+
+  @Override
+  public void run(CommandLine cl) throws IOException {
+    String[] args = cl.getArgs();
+    AlluxioURI srcPath = new AlluxioURI(args[0]);
+    AlluxioURI dstPath = new AlluxioURI(args[1]);
+    List<AlluxioURI> srcPaths = AlluxioShellUtils.getAlluxioURIs(mFileSystem, srcPath);
+    if (srcPaths.size() == 0) {
+      throw new IOException(srcPath.getPath() + " does not exist.");
+    }
+
+    if (srcPath.containsWildcard()) {
+      copyWildcard(srcPaths, dstPath);
+    } else {
+      copy(srcPath, dstPath);
+    }
+  }
+
+  /**
+   * Copies a list of files or directories specified by srcPaths to the destination specified by
+   * dstPath. This method is used when the original source path contains wildcards.
+   *
+   * @param srcPaths a list of files or directories in the Alluxio filesystem
+   * @param dstPath the destination in the Alluxio filesystem
+   * @throws IOException
+   */
+  private void copyWildcard(List<AlluxioURI> srcPaths, AlluxioURI dstPath) throws IOException {
+    URIStatus dstStatus = null;
+    try {
+      dstStatus = mFileSystem.getStatus(dstPath);
+    } catch (FileDoesNotExistException e) {
+      // if the destination does not exist, it will be created
+    } catch (AlluxioException e) {
+      throw new IOException(e.getMessage());
+    }
+
+    if (dstStatus != null && !dstStatus.isFolder()) {
+      throw new IOException(
+          ExceptionMessage.DESTINATION_FILE_CANNOT_EXIST_WITH_WILDCARD_SOURCE.getMessage());
+    }
+    if (dstStatus == null) {
+      try {
+        mFileSystem.createDirectory(dstPath);
```

Comment:
in this code path (`copyWildcard` method) we are copying multiple files onto a directory